RandomizedSVD.C

```diff
@@ -0,0 +1,267 @@
+                                 d_dims[static_cast<unsigned>(rank)], num_cols,
+                                 rank);
+       }
+      }
+      else {
+        int num_transposed_rows = num_cols / d_num_procs;
+        if (num_cols % d_num_procs > d_rank) {
+          num_transposed_rows++;
+        }
+       snapshot_matrix = new Matrix(num_transposed_rows,
+          num_rows, true);
+       snapshot_matrix_transposed = new Matrix(num_transposed_rows,
+          num_rows, true);
+       for (int rank = 0; rank < d_num_procs; ++rank) {
+         gather_block(&snapshot_matrix->item(0, 0), d_samples.get(),
+                                d_istarts[static_cast<unsigned>(rank)] + 1, 1,
+                                d_dims[static_cast<unsigned>(rank)], num_cols,
+                                rank);
+         gather_transposed_block(&snapshot_matrix_transposed->item(0, 0), d_samples.get(),
+                                d_istarts[static_cast<unsigned>(rank)] + 1, 1,
+                                d_dims[static_cast<unsigned>(rank)], num_cols,
+                                rank);
+       }
+      }
+
+     // Create a random matrix of smaller dimension to project the snapshot matrix
+     Matrix* rand_mat = new Matrix(snapshot_matrix->numColumns(), d_subspace_dim, false, true);
+
+     // Project snapshot matrix onto random subspace
+     Matrix* rand_proj = snapshot_matrix->mult(rand_mat);
+
+     // Get QR factorization of random projection
+     int *row_offset = new int[d_num_procs + 1];
+     row_offset[d_num_procs] = rand_proj->numDistributedRows();
+
+     int my_rank;
+     MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
+
+     row_offset[my_rank] = rand_proj->numRows();
+
+     CAROM_VERIFY(MPI_Allgather(MPI_IN_PLACE,
+              1,
+              MPI_INT,
+              row_offset,
+              1,
+              MPI_INT,
+              MPI_COMM_WORLD) == MPI_SUCCESS);
+      for (int i = d_num_procs - 1; i >= 0; i--) {
+        row_offset[i] = row_offset[i + 1] - row_offset[i];
+      }
+
+     CAROM_VERIFY(row_offset[0] == 0);
+
+     SLPK_Matrix slpk_rand_proj;
+
+     d_nprow = d_num_procs;
+     d_npcol = 1;
+     int d_blocksize = row_offset[d_num_procs] / d_num_procs;
+     if (row_offset[d_num_procs] % d_num_procs != 0) d_blocksize += 1;
+       initialize_matrix(&slpk_rand_proj, rand_proj->numDistributedRows(), rand_proj->numColumns(),
+         d_nprow, d_npcol, d_blocksize, d_blocksize);
+     for (int rank = 0; rank < d_num_procs; ++rank) {
+        scatter_block(&slpk_rand_proj, row_offset[rank] + 1,
+                  1, rand_proj->getData(), row_offset[rank + 1] - row_offset[rank],
+                  rand_proj->numColumns(), rank);
+     }
+
+     QRManager QRmgr;
+     qr_init(&QRmgr, &slpk_rand_proj);
+     qrfactorize(&QRmgr);
+
+     // Manipulate lq_A to get elementary household reflectors.
+     Matrix* lq_A = new Matrix(QRmgr.A->mdata, d_subspace_dim,
```

Comment:
Is this used, or is a scalapack routine used instead?

Reply:
pdormqr assumes that the input Q contains the elementary householder reflections. Since the output to pdegpf combines Q and R into one matrix, we need to remove R by zeroing out some values and changing some values to 1, which is done in the following lines.